README.md
# HPE OneView SDK for Chef

## Build Status 

OV Version | 6.20 | 6.10 | 6.00 | 5.60 | 5.50 |
| ------------- |:-------------:| -------------:| -------------:| -------------:| -------------:|
SDK Version/Tag | [v6.2.0](https://github.com/HewlettPackard/oneview-chef/releases/tag/v6.2.0) | [v6.1.0](https://github.com/HewlettPackard/oneview-chef/releases/tag/v6.1.0) | [v6.0.0](https://github.com/HewlettPackard/oneview-chef/releases/tag/v6.0.0) | [v3.8.0](https://github.com/HewlettPackard/oneview-chef/releases/tag/v3.8.0) | [v3.7.0](https://github.com/HewlettPackard/oneview-chef/releases/tag/v3.7.0) |
Build Status | [![Build status](https://api.travis-ci.org/HewlettPackard/oneview-chef.svg?branch=master&status=passed)](https://travis-ci.org/github/HewlettPackard/oneview-chef/builds)| [![Build status](https://api.travis-ci.org/HewlettPackard/oneview-chef.svg?branch=master&status=passed)](https://travis-ci.org/github/HewlettPackard/oneview-chef/builds/767953861)| [![Build status](https://api.travis-ci.org/HewlettPackard/oneview-chef.svg?branch=master&status=passed)](https://travis-ci.org/github/HewlettPackard/oneview-chef/builds/762275325)| [![Build status](https://api.travis-ci.org/HewlettPackard/oneview-chef.svg?branch=master&status=passed)](https://travis-ci.org/github/HewlettPackard/oneview-chef/builds/758660240)| ![Build status](https://api.travis-ci.org/HewlettPackard/oneview-chef.svg?branch=master&status=passed)|

## Introduction

HPE OneView makes it simple to deploy and manage today’s complex hybrid cloud infrastructure. HPE OneView can help you transform your data center to software-defined, and it supports HPE’s broad portfolio of servers, storage, and networking solutions, ensuring the simple and automated management of your hybrid infrastructure. Software-defined intelligence enables a template-driven approach for deploying, provisioning, updating, and integrating compute, storage, and networking infrastructure.

The HPE OneView Chef SDK enables developers to easily build integrations and scalable solutions with HPE OneView and HPE Image Streamer. You can find the latest supported HPE OneView Chef SDK [here](https://github.com/HewlettPackard/oneview-chef/releases/latest)

## What's New

HPE OneView Chef library extends support of the SDK to OneView REST API version 3000 (OneView v6.20)

Please refer to [notes](https://github.com/HewlettPackard/oneview-chef/blob/master/CHANGELOG.md) for more information on the changes , features supported and issues fixed in this version

## Getting Started 

## Requirements
 - Ruby >= 2.4.1 (We recommend using Ruby >= 2.4.1)
 - Chef >= 12.0  (We recommend using Chef >= 13.12 if possible)

## Installation and Configuration
HPE OneView SDK for Chef can be installed from Source and Docker container installation methods. You can either use a docker container which will have the HPE OneView SDK for Chef installed or perform local installation manually.
	
## Installation

## Docker Setup for oneview-chef
The light weight containerized version of the HPE OneView SDK for Chef is available in the [Docker Store](https://hub.docker.com/r/hewlettpackardenterprise/hpe-oneview-sdk-for-chef). The Docker Store image tag consist of two sections: <sdk_version-OV_version>

```bash
# Download and store a local copy of oneview-chef and
# use it as a Docker image.
$ docker pull hewlettpackardenterprise/hpe-oneview-sdk-for-chef:v6.2.0-OV6.2
# Run docker commands below given, which  will in turn create
# a sh session where you can create files, issue commands and execute the recipes.
$ docker run -it hewlettpackardenterprise/hpe-oneview-sdk-for-chef:v6.2.0-OV6.2 /bin/sh
```

## Local Setup for oneview-chef
Chef SDK dependencies can be installed by bundler.

```bash
# Create a folder cookbooks and clone the chef repo
$ mkdir cookbooks
$ cd cookbooks
$ git clone https://github.com/chef-boneyard/compat_resource
# Clone Chef repo as oneview
$ git clone https://github.com/HewlettPackard/oneview-chef oneview
$ cd oneview
```

### Local installation requires the gem in your Gemfile:
  ```ruby
  gem 'oneview-sdk', '~> 6.2.0'
  ```
### Install Chef Sdk dependencies from Gemfile

```bash
$ bundle install
```

## Usage
The cookbook 'metadata' is not intended to include any recipes instead specifies a dependency which is used by another cookbooks.

```ruby
# my_cookbook/metadata.rb
depends 'oneview', '~> 6.2.0'
```

## Credentials
In order to manage HPE OneView and HPE Synergy Image Streamer resources, you will need to provide authentication credentials. There are 2 ways to do this:

### Environment variables:
  - For HPE OneView: 
    ```bash
    $export ONEVIEWSDK_URL=<ov-endpoint>
    $export ONEVIEWSDK_USER=<ov-user>
    $export ONEVIEWSDK_PASSWORD=<ov-password>
    $export ONEVIEWSDK_SSL_EANBLED=false
    ```

  - For HPE Synergy Image Streamer: 
    ```bash
    $export I3S_URL=<i3s-endpoint>
    ```

### API version
When using the resources a API version will be selected to interact with the resources in each HPE OneView correct API versions. To select the desired one, you may use one of the following methods:

1. Set the resource property `api_version`. See the [Resource Properties](#resource-properties) section below.
2. Set the client parameter `api_version`. If this parameter is set, it will select the required API version based on the client. Notice if you choose to pass the client as an OneviewSDK object, it will have, by default, the api_version set, even if you do not directly specify it.
3. If none of the previous alternatives are set, it defaults to the `node['oneview']['api_version']`. See the [Atributes](#attributes).

Be aware of the precedence of these methods! The higher priority goes to setting the resource property, followed by the client parameter, and at last the node value as the default, i.e. *Property > Client > Attribute*. (e.g. If you set the resource property `api_version` to *200*, and set the client parameter `api_version` to *300*, it will use the module **API200**, since the resource property takes precedence over the client parameter)

## Attributes
 - `node['oneview']['ruby_sdk_version']` - Set which version of the SDK to install and use. Defaults to `'~> 4.1'`
 - `node['oneview']['save_resource_info']` - Save resource info to a node attribute? Defaults to `['uri']`. Possible values/types:
   - `true` - Save all info (Merged hash of OneView info and Chef resource properties). Warning: Resource credentials will be saved if specified.
   - `false` - Do not save any info.
   - `Array` - i.e. `['uri', 'status', 'created_at']` Save a subset of specified attributes.
 - `node['oneview']['api_version']` - When looking for a matching Chef resource provider class, this version will be used as default. Defaults to `200`.
 - `node['oneview']['api_variant']` - When looking for a matching Chef resource provider class, this variant will be used as default. Defaults to `C7000`.

### Client Property
Explicitly pass in the `client` property to each resource. This takes precedence over environment variables and allows you to set more client properties. This also allows you to get these credentials from other sources like encrypted databags, Vault, etc.

```ruby
my_client = {
  url: 'https://example.com',
  user: 'username',
  password: 'password',
  api_version: 3000
}
```

HPE Synergy Image Streamer access token is the same as the HPE OneView associated appliance, so most of its credentials you may get from the HPE OneView.

## Resources

### Resource Properties
The following are the standard properties available for all resources. Some resources have additional properties or small differences.

```ruby
oneview_ethernet_network 'Eth1' do
  client my_client
  api_version 3000
  api_variant 'Synergy'
  data(
    vlanId: 1001,
    purpose: 'General',
    smartLink: false,
    privateNetwork: false,
    bandwidth: {
      typicalBandwidth: 2000,
      maximumBandwidth: 9000
    }
  )
  action :create_if_missing
end
```

## Examples
The examples provided here will help you to understand the detailed usage of each resource in Chef Sdk. It allows users to easily create, update, query and delete resources.

You can find the resource examples [here](https://github.com/HewlettPackard/oneview-chef/blob/master/EXAMPLES.md). This document provides the list of attributes and the methods being used in each resource to perform CRUD operations on the oneview appliance.
 - attributes: client, data, scopes, etc.
 - methods: :create, :delete, :delete_bulk

## Getting Help 

Are you running into a road block? Have an issue with unexpected bahriov? Feel free to open a new issue on the [issue tracker](https://github.com/HewlettPackard/oneview-chef/issues)

For more information on how to open a new issue refer to [How can I get help & support](https://github.com/HewlettPackard/oneview-chef/wiki#getting-help---how-can-i-get-help—support)

## License
This project is licensed under the Apache 2.0 license. Please see [LICENSE](LICENSE) for more info.

## Contributing and feature requests
**Contributing:** You know the drill. Fork it, branch it, change it, commit it, and pull-request it.
We are passionate about improving this project, and glad to accept help to make it better. However, keep the following in mind:

 - You must sign a Contributor License Agreement first. Contact one of the authors (from Hewlett Packard Enterprise) for details and the CLA.
 - We reserve the right to reject changes that we feel do not fit the scope of this project, so for feature additions, please open an issue to discuss your ideas before doing the work.

**Feature Requests:** If you have a need that is not met by the current implementation, please let us know (via a new issue).
This feedback is crucial for us to deliver a useful product. Do not assume we have already thought of everything, because we assure you that is not the case.

## Testing
 - Style (Rubocop & Foodcritic): `$ rake style`
 - Unit: `$ rake unit`
 - Run all tests: `$ rake test`
 - Optional: Start guard to run tests automatically on file changes: `$ bundle exec guard`

 For more information please refer to the [Testing guidelines](TESTING.md).

## Contributors

You can find the list of contributors [here](https://github.com/HewlettPackard/oneview-chef/graphs/contributors)

## Additional Resources 

**HPE OneView Documentation**

[HPE OneView Release Notes](http://hpe.com/info/OneView/docs)

[HPE OneView Support Matrix](http://hpe.com/info/OneView/docs)

[HPE OneView Installation Guide](http://hpe.com/info/OneView/docs)

[HPE OneView User Guide](http://hpe.com/info/OneView/docs)

[HPE OneView Online Help](http://hpe.com/info/OneView/docs)

[HPE OneView REST API Reference](http://hpe.com/info/OneView/docs)

[HPE OneView Firmware Management White Paper](http://hpe.com/info/OneView/docs)

[HPE OneView Deployment and Management White Paper](http://hpe.com/info/OneView/docs)


**HPE OneView Community**

[HPE OneView Community Forums](http://hpe.com/info/oneviewcommunity)

Learn more about HPE OneView at [hpe.com/info/oneview](https://hpe.com/info/oneview)


CHANGELOG.md
## 6.2.0
### Notes
- This release extends support of the SDK to OneView REST API version 3000 (OneView v6.20).

### Oneview Features supported
- oneview_connection_template
- oneview_enclosure
- oneview_enclosure_group
- oneview_ethernet_network
- oneview_fc_network
- oneview_fcoe_network
- oneview_firmware
- oneview_hypervisor_cluster_profile
- oneview_hypervisor_manager
- oneview_interconnect
- oneview_logical_enclosure
- oneview_logical_interconnect
- oneview_logical_interconnect_group
- oneview_network_set
- oneview_scope
- oneview_server_certificate
- oneview_server_hardware
- oneview_server_hardware_type
- oneview_server_profile
- oneview_server_profile_template
- oneview_storage_pool
- oneview_storage_system
- oneview_uplink_set
- oneview_volume
- oneview_volume_attachment
- oneview_volume_template

## 6.1.0
### Notes
- This release extends support of the SDK to OneView REST API version 2800 (OneView v6.10) and ImageStreamer REST API version 2020 (I3S v6.10).

### Oneview Features supported
- oneview_connection_template
- oneview_enclosure
- oneview_enclosure_group
- oneview_ethernet_network
- oneview_fc_network
- oneview_fcoe_network
- oneview_firmware
- oneview_hypervisor_cluster_profile
- oneview_hypervisor_manager
- oneview_interconnect
- oneview_logical_enclosure
- oneview_logical_interconnect
- oneview_logical_interconnect_group
- oneview_network_set
- oneview_scope
- oneview_server_certificate
- oneview_server_hardware
- oneview_server_hardware_type
- oneview_server_profile
- oneview_server_profile_template
- oneview_storage_pool
- oneview_storage_system
- oneview_uplink_set
- oneview_volume
- oneview_volume_attachment
- oneview_volume_template

### Image Streamer Features supported
- image_streamer_artifact_bundle
- image_streamer_deployment_plan

## 6.0.0
### Notes
- This release extends support of the SDK to OneView REST API version 2600 (OneView v6.00) and ImageStreamer REST API version 2010 (I3S v6.00).

### Oneview Features supported
- oneview_connection_template
- oneview_enclosure
- oneview_enclosure_group
- oneview_ethernet_network
- oneview_fc_network
- oneview_fcoe_network
- oneview_firmware
- oneview_hypervisor_cluster_profile
- oneview_hypervisor_manager
- oneview_interconnect
- oneview_logical_enclosure
- oneview_logical_interconnect
- oneview_logical_interconnect_group
- oneview_network_set
- oneview_scope
- oneview_server_certificate
- oneview_server_hardware
- oneview_server_hardware_type
- oneview_server_profile
- oneview_server_profile_template
- oneview_storage_pool
- oneview_storage_system
- oneview_uplink_set
- oneview_volume
- oneview_volume_attachment
- oneview_volume_template

### Image Streamer Features supported
- image_streamer_artifact_bundle
- image_streamer_deployment_plan

## 3.8.0
### Notes
- This release extends support of the SDK to OneView REST API version 2400 (OneView v5.60).

### Oneview Features supported
- oneview_connection_template
- oneview_enclosure
- oneview_enclosure_group
- oneview_ethernet_network
- oneview_fc_network
- oneview_fcoe_network
- oneview_hypervisor_cluster_profile
- oneview_hypervisor_manager
- oneview_interconnect
- oneview_logical_enclosure
- oneview_logical_interconnect
- oneview_logical_interconnect_group
- oneview_network_set
- oneview_scope
- oneview_server_certificate
- oneview_server_hardware
- oneview_server_hardware_type
- oneview_server_profile
- oneview_server_profile_template
- oneview_storage_pool
- oneview_storage_system
- oneview_uplink_set
- oneview_volume
- oneview_volume_attachment
- oneview_volume_template

## 3.7.0
### Notes
- This release extends support of the SDK to OneView REST API version 2200 (OneView v5.50) and ImageStreamer REST API version 2000 (I3S v5.40).

### Oneview Features supported
- oneview_connection_template
- oneview_enclosure
- oneview_enclosure_group
- oneview_ethernet_network
- oneview_fc_network
- oneview_fcoe_network
- oneview_hypervisor_cluster_profile
- oneview_hypervisor_manager
- oneview_interconnect
- oneview_logical_enclosure
- oneview_logical_interconnect
- oneview_logical_interconnect_group
- oneview_network_set
- oneview_scope
- oneview_server_certificate
- oneview_server_hardware
- oneview_server_hardware_type
- oneview_server_profile
- oneview_server_profile_template
- oneview_storage_pool
- oneview_storage_system
- oneview_uplink_set
- oneview_volume
- oneview_volume_attachment
- oneview_volume_template

### Image Streamer Features supported
- image_streamer_artifact_bundle
- image_streamer_deployment_plan

## 3.6.0

## Enhancements:
- Enhanced the method to set the default API version to appliance's max API version instead of hardcoded 200.

### Notes
- This release extends support of the SDK to OneView REST API version 2000 (OneView v5.40).

### Oneview Features supported
- oneview_connection_template
- oneview_enclosure
- oneview_enclosure_group
- oneview_ethernet_network
- oneview_fc_network
- oneview_fcoe_network
- oneview_hypervisor_cluster_profile
- oneview_hypervisor_manager
- oneview_interconnect
- oneview_logical_enclosure
- oneview_logical_interconnect
- oneview_logical_interconnect_group
- oneview_network_set
- oneview_scope
- oneview_server_certificate
- oneview_server_hardware
- oneview_server_hardware_type
- oneview_server_profile
- oneview_server_profile_template
- oneview_storage_pool
- oneview_storage_system
- oneview_uplink_set
- oneview_volume
- oneview_volume_attachment
- oneview_volume_template

## 3.5.0
### Notes
- This release extends support of the SDK to OneView REST API version 1800 (OneView v5.30).

### Oneview Features supported
- oneview_connection_template
- oneview_enclosure
- oneview_enclosure_group
- oneview_ethernet_network
- oneview_fc_network
- oneview_fcoe_network
- oneview_hypervisor_cluster_profile
- oneview_hypervisor_manager
- oneview_interconnect
- oneview_logical_enclosure
- oneview_logical_interconnect
- oneview_logical_interconnect_group
- oneview_network_set
- oneview_scope
- oneview_server_certificate

examples/fabric.rb
# (c) Copyright 2017 Hewlett Packard Enterprise Development LP
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.

my_client = {
  url: ENV['ONEVIEWSDK_URL'],
  user: ENV['ONEVIEWSDK_USER'],
  password: ENV['ONEVIEWSDK_PASSWORD']
}

# Updates the DefaultFabric fabric reserved vlan range if the values do not match
# Only available for API300 with Synergy
oneview_fabric 'DefaultFabric' do
  client my_client
  api_version 300
  api_variant 'Synergy'
  reserved_vlan_range(
    'start' => 3000,
    'length' => 120
  )
end
